devito/types/sparse.py

```diff
@@ -931,23 +936,99 @@ class PrecomputedSparseFunction(AbstractSparseFunction):
-    is_PrecomputedSparseFunction = True
-
     _sub_functions = ('gridpoints', 'interpolation_coeffs')
 
+    __rkwargs__ = (AbstractSparseFunction.__rkwargs__ +
+                   ('r', 'coordinates_data', 'gridpoints_data',
+                    'interpolation_coeffs_data'))
+
     def __init_finalize__(self, *args, **kwargs):
-        super(PrecomputedSparseFunction, self).__init_finalize__(*args, **kwargs)
+        super().__init_finalize__(*args, **kwargs)
 
         # Grid points per sparse point (2 in the case of bilinear and trilinear)
         r = kwargs.get('r')
+        if not is_integer(r):
+            raise TypeError('Need `r` int argument')
+        if r <= 0:
+            raise ValueError('`r` must be > 0')
+        self._r = r
+
+        coordinates = kwargs.get('coordinates', kwargs.get('coordinates_data'))
+        if isinstance(coordinates, SubFunction):
+            self._coordinates = coordinates
+        elif isinstance(coordinates, Iterable):
+            coordinates_data = coordinates
+            dimensions = (self.indices[self._sparse_position], Dimension(name='d'))
+            shape = (self.npoint, self.grid.dim)
+            self._coordinates = SubFunction(
+                name='%s_coords' % self.name, parent=self, dtype=self.dtype,
+                dimensions=dimensions, shape=shape, space_order=0,
+                initializer=coordinates_data, alias=self.alias,
+                distributor=self._distributor
+            )
+        elif coordinates is None:
+            # Unlike `gridpoints` or `interpolation_coeffs`, not strictly necessary
+            self._coordinates = None
+        else:
+            raise ValueError("`coordinates` must be either SubFunction or iterable "
+                             "(e.g., list, np.ndnarray)")
+
         gridpoints = kwargs.get('gridpoints')
-        interpolation_coeffs = kwargs.get('interpolation_coeffs')
+        gridpoints = kwargs.get('gridpoints', kwargs.get('gridpoints_data'))
+        if isinstance(gridpoints, SubFunction):
+            self._gridpoints = gridpoints
+        elif isinstance(gridpoints, Iterable):
+            gridpoints_data = gridpoints
+            dimensions = (self.indices[self._sparse_position], Dimension(name='d'))
+            shape = (self.npoint, self.grid.dim)
+            self._gridpoints = SubFunction(
+                name="%s_gridpoints" % self.name, parent=self, dtype=np.int32,
+                dimensions=dimensions, shape=shape, space_order=0,
+                initializer=gridpoints_data, alias=self.alias,
+                distributor=self._distributor,
+            )
+        else:
+            raise ValueError("`gridpoints` must be either SubFunction or iterable "
+                             "(e.g., list, np.ndnarray)")
+
+        interpolation_coeffs = kwargs.get('interpolation_coeffs',
+                                          kwargs.get('interpolation_coeffs_data'))
+        if isinstance(interpolation_coeffs, SubFunction):
```

Comment:
I think could be cleanedup wiht `_coordinate_setup(name=gripoint/coeffs/cords/...` that's a lot of repetition to maintain.